### PR TITLE
Be more careful when selecting values and vectors in svds

### LINKED ISSUE #335: `svds` returns non-orthogonal singular vectors when there are degenerate singular values
A colleague of mine spotted a bug with the following:

``` jl
function bugdemo(D)
        M = diagm(D)
        U, S, V = svds(M, nsv=4)
        C = U' * M * V
        C[abs(C) .< 1e-12] = 0.
        println(diagm(S))  # These two should
        println(C)         # be the same.

        ide = U' * U
        ide[abs(ide) .< 1e-12] = 0
        println(ide)
end

D = rand(5)
D[5] = 0
bugdemo(D)

println()
D[2] = D[1]
bugdemo(D)
```

That prints out for instance

```
[0.9425868903202645 0.0 0.0 0.0
 0.0 0.7436506177105782 0.0 0.0
 0.0 0.0 0.3137218131235404 0.0
 0.0 0.0 0.0 0.1446005101113296]
[0.942586890320264 0.0 0.0 0.0
 0.0 0.7436506177105787 0.0 0.0
 0.0 0.0 0.3137218131235404 0.0
 0.0 0.0 0.0 0.14460051011132963]
[0.9999999999999993 0.0 0.0 0.0
 0.0 1.0000000000000009 0.0 0.0
 0.0 0.0 1.0 0.0
 0.0 0.0 0.0 0.9999999999999993]

[0.9425868903202645 0.0 0.0 0.0
 0.0 0.9425868903202641 0.0 0.0
 0.0 0.0 0.7436506177105784 0.0
 0.0 0.0 0.0 0.31372181312354036]
[0.942586890320264 -0.8015220638016495 0.0 0.0
 -0.8015220638016496 0.9425868903202647 0.0 0.0
 0.0 0.0 0.7436506177105782 0.0
 0.0 0.0 0.0 0.31372181312354047]
[0.9999999999999996 -0.8503428936183435 0.0 0.0
 -0.8503428936183435 1.0000000000000007 0.0 0.0
 0.0 0.0 0.9999999999999984 0.0
 0.0 0.0 0.0 1.0000000000000004]
```

So in the first case there are no degenerate singular values and everything goes fine: After doing an SVD of M, we can multiply M with U and S and get a diagonal matrix with the singular values on the diagonal. The columns of U are orthogonal.

However, if we set `D[2] = D[1]` so that there's degeneracy, the same thing fails. The singular values come out right, but the singular vectors don't: You can see that the block of C corresponding to the degenerate singular values is not diagonal and the columns of U are not orthogonal.

I'm guessing that this is related to the non-uniqueness of singular vectors corresponding to degenerate singular values, but I haven't looked at the code for `svds`. Note that if you run that code a few times, sometimes things work out correctly by luck even in the latter case.

### LINKED ISSUE #16608: `svds` returns non-orthogonal singular vectors when there are degenerate singular values
A colleague of mine spotted a bug with the following:

``` jl
function bugdemo(D)
        M = diagm(D)
        U, S, V = svds(M, nsv=4)
        C = U' * M * V
        C[abs(C) .< 1e-12] = 0.
        println(diagm(S))  # These two should
        println(C)         # be the same.

        ide = U' * U
        ide[abs(ide) .< 1e-12] = 0
        println(ide)
end

D = rand(5)
D[5] = 0
bugdemo(D)

println()
D[2] = D[1]
bugdemo(D)
```

That prints out for instance

```
[0.9425868903202645 0.0 0.0 0.0
 0.0 0.7436506177105782 0.0 0.0
 0.0 0.0 0.3137218131235404 0.0
 0.0 0.0 0.0 0.1446005101113296]
[0.942586890320264 0.0 0.0 0.0
 0.0 0.7436506177105787 0.0 0.0
 0.0 0.0 0.3137218131235404 0.0
 0.0 0.0 0.0 0.14460051011132963]
[0.9999999999999993 0.0 0.0 0.0
 0.0 1.0000000000000009 0.0 0.0
 0.0 0.0 1.0 0.0
 0.0 0.0 0.0 0.9999999999999993]

[0.9425868903202645 0.0 0.0 0.0
 0.0 0.9425868903202641 0.0 0.0
 0.0 0.0 0.7436506177105784 0.0
 0.0 0.0 0.0 0.31372181312354036]
[0.942586890320264 -0.8015220638016495 0.0 0.0
 -0.8015220638016496 0.9425868903202647 0.0 0.0
 0.0 0.0 0.7436506177105782 0.0
 0.0 0.0 0.0 0.31372181312354047]
[0.9999999999999996 -0.8503428936183435 0.0 0.0
 -0.8503428936183435 1.0000000000000007 0.0 0.0
 0.0 0.0 0.9999999999999984 0.0
 0.0 0.0 0.0 1.0000000000000004]
```

So in the first case there are no degenerate singular values and everything goes fine: After doing an SVD of M, we can multiply M with U and S and get a diagonal matrix with the singular values on the diagonal. The columns of U are orthogonal.

However, if we set `D[2] = D[1]` so that there's degeneracy, the same thing fails. The singular values come out right, but the singular vectors don't: You can see that the block of C corresponding to the degenerate singular values is not diagonal and the columns of U are not orthogonal.

I'm guessing that this is related to the non-uniqueness of singular vectors corresponding to degenerate singular values, but I haven't looked at the code for `svds`. Note that if you run that code a few times, sometimes things work out correctly by luck even in the latter case.

### PR DESCRIPTION
**Pull request by andreasnoack**
Fixes JuliaLang/LinearAlgebra.jl#335

Make "Symmetric generalized with singular B" a testset

### CODE REVIEW

**Comment by tkelman** [comment]
doesn't this make the size of the returned output less predictable?

**Comment by andreasnoack** [comment]
I don't think so. The eigenvalues should always come pairs of one positive and one negative. When a singular value is repeated, the ordering of the eigenvalues might be distorted but the number of positive values should remain the same.

**Comment by tkelman** [comment]
could we add a consistency check for that then?

**Comment by andreasnoack** [comment]
When trying to add a check, I realized that the old version of extracting the values and vectors was actually wrong since it used the subspace size and not `nsv` to pick the values and vectors. Instead of asking for `which=:LM` with `nev=2nsv`, I've change the call to `eigs` to `which=:LR` with `nev=nsv`. That should avoid the problem of finding the which right values and vectors among the eigenpairs and should also be faster because only half the values are computed.

File: test/linalg/arnoldi.jl [hunk 1]
```diff
@@ -3,78 +3,83 @@
...
+        bsym  = b' + b
+        bpd   = b'*b
 
-            (d,v) = eigs(apd, nev=3)
-            @test apd*v[:,3] ≈ d[3]*v[:,3]
-            @test eigs(apd; nev=1, sigma=d[3])[1][1] ≈ d[3]
+        (d,v) = eigs(a, nev=3)
+        @test a*v[:,2] ≈ d[2]*v[:,2]
+        @test norm(v) > testtol # eigenvectors cannot be null vectors
+        # (d,v) = eigs(a, b, nev=3, tol=1e-8) # not handled yet
```
Comment (tkelman) [inline test/linalg/arnoldi.jl]: could you use test_broken here?

Comment (andreasnoack) [inline test/linalg/arnoldi.jl]: I can do that.

**Comment by andreasnoack** [comment]
I'm puzzled by what is going on here. If I run just the arpack tests then they pass but they fail with `testall`. This is surprising because I'm setting the seed so all the random values should be the unaffected. What else could cause this? Rounding modes?

**Comment by andreasnoack** [comment]
Superseded by #21701

**Pull request closed without merging**